### DISCUSSION TITLE
issue: setValue in onSuccess of react-query has an unknown behavior

### DISCUSSION THREAD

**Discussion by fardad-dev** [react-hook-form/react-hook-form#7767] [Q&A] [answered]
### Version Number

7.26.0

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-js-forked-bji03?file=/src/App.js

### Steps to reproduce

1. Click on clear

### Expected behaviour

Expected to clear Selected item of dropdown

What happen:
In onClick of clear button category was set to undefined then value of category is undefined when you calling getValues but Controller field has the value which was set in onSuccess of useQuery.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
please provide some more detail, I don't really get the use case in here. 

eg: 

1. steps1 click on clear 
2. what to expect? in the console?

**Comment by amin-gharouni** [comment]
I have the same issue
I have a api that set some data to my form as default values but the problem is if user wants to clear the value, my component won't update its value and keeps the previous value 
it happens if I set my form value to undefined

**Comment by fardad-dev** [comment]
> please provide some more detail, I don't really get the use case in here.
> 
> eg:
> 
> 1. steps1 click on clear
> 2. what to expect? in the console?

I changed code, now logging field.value.
1- steps1 click on clear
2- I expected to log `controllerValue undefined` but `controllerValue 1` was logged

**Comment by bluebill1049** [comment]
https://react-hook-form.com/api/usecontroller

> You need to either set defaultValue at the field-level or call useForm with defaultValues. If both are set, defaultValue will be used.

https://react-hook-form.com/api/useform

![Screen Shot 2022-02-07 at 7 49 02 pm](https://user-images.githubusercontent.com/10513364/152754792-3de0eb49-c475-4ea2-ac0c-07ac4afacd7d.png)

please supply a none `undefined` value according to the doc.

**Comment by Moshyfawn** [comment]
An expected behavior.
See #7657 for a discussion on this behavior.

**Comment by Moshyfawn** [comment]
Also, pls, take a note of that RQ loves to re-fetch data a lot by default. It's stale data's gonna be re-fetched on window focus and in a lot of other circumstance that you can read about [here](https://react-query.tanstack.com/guides/important-defaults). That said, setting a value _(`setValue`)_ in `onSuccess` of `useQuery` will result in a frequent update of your form field value, so if you try clearing you input and, say, an alert will pop up, closing it will result in window focus event and `onSuccess` callback fired _(if the query was successful)_ which will lead to your just cleared value being set to what you're setting it to in `onSuccess`

**Comment by fardad-dev** [comment]
> 

Your all right, issue is not about this.

**Comment by fardad-dev** [comment]
> https://react-hook-form.com/api/usecontroller
> 
> > You need to either set defaultValue at the field-level or call useForm with defaultValues. If both are set, defaultValue will be used.
> 
> https://react-hook-form.com/api/useform
> 
> ![Screen Shot 2022-02-07 at 7 49 02 pm](https://user-images.githubusercontent.com/10513364/152754792-3de0eb49-c475-4ea2-ac0c-07ac4afacd7d.png)
> 
> please supply a none `undefined` value according to the doc.

This so hard to change typeof value our components has type like (number | undefined) we can't change to another types. what's your suggest?

**Comment by fardad-dev** [comment]
I have another question. why that was work with setTimout?

**Comment by fardad-dev** [comment]
Also when you don't set default value setValue to undefined not working

**Comment by bluebill1049** [comment]
> This so hard to change typeof value our components has type like (number | undefined) we can't change to another types. what's your suggest?

I think it's better to follow native input with default as an empty string instead `undefined` as we stated in the doc, `undefined` is not ideal even for a controlled component.

**Comment by Moshyfawn** [comment]
> Also when you don't set default value set to undefined not working

By not setting your `defaultValue`, it's `undefined`
> This so hard to change typeof value our components has type like (number | undefined) we can't change to another types. what's your suggest?

There's no recommendation other than to have `null` as your default value. Antd Select treats `""` _(empty string)_ as a non-empty value, so your best bet is to pass `null` to `setValue` to reset your input.

Also, there's [a `resetField` form method](https://react-hook-form.com/api/useform/resetfield) that's more suitable for for clearing inputs to their default value. It'll reset your input to the initially set default value on `useForm`
```js
const { resetField } = useForm({
  defaultValues: {
    category: null
  } 
})

<button onClick={() => resetField("category")}>Clear</button>
```

**Comment by bluebill1049** [comment]
> I have another question. why that was work with setTimout?

I would like to investigate why, but I don't think there is too much value as we don't support `undefined` in the first place. My guess is setTimeout will make react scheduling on the next state update, potentially having a value in place.

**Comment by bluebill1049** [comment]
> Also when you don't set default value setValue to undefined not working

if you are working aginst this library, there are probably more expected behavior which you will experience.

**Comment by Moshyfawn** [comment]
> @bluebill1049 I think it's better to follow native input with default as an empty string instead undefined

Antd Select treats `""` _(empty string)_ as a non-empty value. I guess, `null` is the way to go for this input

**Comment by bluebill1049** [comment]
thank you @Moshyfawn Yea, `null` would be the preference here.

**Comment by Moshyfawn** [comment, marked as answer]
At the end of the day, [here's how](https://codesandbox.io/s/react-hook-form-js-forked-jdjhn?file=/src/App.js) I'd go about the form that you have
- Set the default `category` field value as `null`
- Use [the `resetField` form method](https://react-hook-form.com/api/useform/resetfield) **inside the `useEffect`** to prevent setting `category` field value from updating every query re-fetch. Set the first data option `id` as the select's `defaultValue`, because it seems like this is what you're trying to do in your example.
```js
useEffect(() => {
  if (data) {
    resetField("category", { defaultValue: data[0]?.id });
  }
}, [resetField, data]);
``` 
- Call [the `resetField` form method](https://react-hook-form.com/api/useform/resetfield) to clear the select input to it's default value. It will result in the select's value being the first data option `id`
```js
<button onClick={() => resetField("category")}>Clear</button>
```

This ensures that
- You don't override the select's value on query re-fetch
- Sets select's value to a valid default value _(`null` initially)_ that doesn't go against React's rules of controlled vs uncontrolled inputs
- As you set the field's value to the first option of the fetched data, clearing the input won't go against it and have the default value of that first option.

**Reply by bluebill1049** [reply]
I think it makes sense to make the above the answer to this question.

**Reply by fardad-dev** [reply]
`null` have an issue with default value of components, for example  I have many components like this
```js
({value="default"}) => <Somtthing />
```
If I use `null` instead of  `undefined`, component will behavior differently.

**Reply by Moshyfawn** [reply]
Well, it looks like you're providing a non `undefined` to your component's `value` after all, so not setting a `defaultValue` (aka providing `undefined`) should be fine in theory.
Now, you still can't provide `undefined` to `setValue` as it's reserved by RHF as a fallback. It's gonna be like that at least in the nearest RHF versions, so, you can consider it an API limitation for you

**Comment by Moshyfawn** [comment]
I'd say, this is as much issue resolution as we can provide from the RHF stand point.

The `undefined` value in React has a meaning to determine whether or not the input is controlled and RHF `Controller` API is expected to be used with the controlled components by design.

If you have any more thoughts on the matter, you're welcome to continue this as a discussion
